### ISSUE TITLE
handle Parental Release in Reminder email

### ISSUE THREAD

**Issue by mscholtz** [washingtontrails/vms#167]
in the reminder email (https://cs9.salesforce.com/00XK0000000Mg7E) if the Parental Release (https://cs9.salesforce.com/a2pK0000000dAgj) is specified as a required document, it is listed as "Please Compelte the Parental Release", but without a link to allow doing so:

![image](https://cloud.githubusercontent.com/assets/931031/8490794/feae2fe2-20e3-11e5-968a-8e9519affe02.png)

Parental release is collected via DocuSign.  The template expects a link in the FormAssembly link field, which we do have for the Medical and Backcountry forms, but don't have for this one.

We should either 
a) figure out how to provide a docusign link in the email to allow completing the form, or
b) use different words, to not lead vols to expect a way to complete the form via the email.  perhaps redirect them back to the documents tab in VMS?

**Comment by mscholtz** [comment]
@WaBirder57 this seems fairly high priority to me, but I'll let you set the priority

**Comment by davisagli** [comment]
Hmm. There isn't a direct URL link to signing the release in docusign, because there's a multi-step process to set up the envelope via the docusign API that needs to happen first. We can link to [pyramid work party URL]/sign-document?r=[registration id]&document=[required document id] which will do those steps and then redirect to docusign...but it currently only works if visited while logged in.

We also need to make sure it only mentions the parental release in the email if the registrant is under 18.

**Comment by mscholtz** [comment]
@davisagli  So what happens on the documents tab for an unauthenticated registrant?  Do they just need to download the form and fill it out on paper?

re you 2nd concern: this document should (I think) only be set (in the WP Type) to required for youth-type WP's.  Right, @WaBirder57 ?  In that case, does every participant need it?  Or are there some cases where we'll have a mix of youth and adults and we want to require this only for the youth?

**Comment by davisagli** [comment]
Right, unauthenticated users only see the links to the printable forms (assuming one is configured on the Required Document in Salesforce).

In the pyramid app we're already only showing the parental release for registrations with "under 18" checked. I was assuming that it would be added as a required document for all work party types that allow youth registration, but that it doesn't make sense to show for adults.

**Comment by mscholtz** [comment]
Charlie, I think we need you to weigh in here on both of these issues.
What should this email show when it's an unauthenticated registrant and
they're missing forms?  And what about the parental release (see my
question earlier).

Thanks!


On Wed, Jul 8, 2015 at 11:14 AM, David Glick <notifications@github.com>
wrote:

> Right, unauthenticated users only see the links to the printable forms
> (assuming one is configured on the Required Document in Salesforce).
>
> In the pyramid app we're already only showing the parental release for
> registrations with "under 18" checked. I was assuming that it would be
> added as a required document for all work party types that allow youth
> registration, but that it doesn't make sense to show for adults.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/washingtontrails/vms/issues/167#issuecomment-119682851>
> .
>



-- 

Matthew Scholtz
Percolator Consulting
(206) 289-0816 x3

http://percolatorconsulting.com/

**Comment by WaBirder57** [comment]
For DocuSign we use the "Powerform" feature.  That is a URL link to a "blank" form that can be shared with the person who needs to complete it.  That starts the workflow for getting that document signed.  Here are the links to the Parental Release form in either URL or HTML syntax:

URL:
 	https://na2.docusign.net/Member/PowerFormSigning.aspx?PowerFormId=ca2658d2-36e4-4fd1-bfac-c29566e7f313

HTML:
 	<a href= 'https://na2.docusign.net/Member/PowerFormSigning.aspx?PowerFormId=ca2658d2-36e4-4fd1-bfac-c29566e7f313'/>Parental Release</a>

**Comment by WaBirder57** [comment]
I think this DocuSign Powerform should be used for either unauthenticated or authenticated users.  This way we only have one "form" that needs to be updated if it changes.

**Comment by WaBirder57** [comment]
So any time there is a youth they will need a parental release form, even if they show up at the trail head with their parent(s).  Ideally we would get this ahead of time via Docusign.

**Comment by davisagli** [comment]
Hmm, I wasn't aware of the Powerform feature.

I don't think it really helps though. If the user isn't logged in, we don't know their contact ID in Salesforce, so we can't make sure that the completed form gets attached to their contact in Salesforce.

It's also not clear to me if we can pre-fill fields (i.e. work party name and date) when using a Powerform like we can when creating an envelope via the Docusign API.

**Comment by davisagli** [comment]
Sorry, I was forgetting the context: when sending the notification from Salesforce, we would have the contact id of everyone in the party. So if it's possible to pass custom field values to a Powerform URL, we could do that. I guess we should experiment and see if that's possible.

**Comment by mscholtz** [comment]
@WaBirder57 I know you've used Powerform in the past, but I  thought the reason we were using the SF-Docusign integration for this stuff is so that results can end up logged in SF.  Do you know whether that's possible w/ Powerform?  (if so, why haven't we been using it all along?)

**Comment by davisagli** [comment]
@mscholtz the Docusign-Salesforce integration ("Docusign Connect") creates records in SF regardless of what way the envelope was created in Docusign -- one "Docusign Status" record holding the document and its status, and a "Docusign Recipient Status" record for each signer. The latter can be linked to a contact IF the envelope has a field containing a contact id and that field is mapped to a field on the Docusign Recipient Status object. In the case of parental releases I also created a custom "Youth Contact Id" field on the envelope and mapped that to a field on the Docusign Status, so that the youth's contact gets linked to the signed document even though they aren't the one signing. I'm not sure if we can pass either of these contact ids using the Powerform URL (maybe we can, but it'll take a few hours of experimentation to find out, and the answer may be no).

An alternative to using a Powerform URL would be to relax the login restriction so that the pyramid URL that creates a Docusign envelope and starts the signing process via the API is accessible to unauthenticated users -- the email notification could then link to that URL for each youth in the party who needs a parental release signed, passing the registration id and document id. This would be a new story because during our initial story creation we were planning to handle this case by directing the volunteer to the Documents tab on the work party's pyramid page, where they can access a printable copy of the form.

**Comment by mscholtz** [comment]
Gotcha, thanks @davisagli 

@WaBirder57  can you do the digging to figure out if we can pass the required info to powerform on the URL?  You're most familiar with it...

**Comment by davisagli** [comment]
@mscholtz this documentation is the most info I could find on passing fields to the Powerform URL: https://www.docusign.com/support/classic/documentation/cdse-powerforms/powerforms/populating-fields-web-powerform

But it's not clear if it supports the specific fields we need to pass, or what exactly they need to be called. Hence why I said a few hours of experimentation.

**Comment by davisagli** [comment]
Do either of you see any problem with my suggestion of linking to the pyramid URL which creates an envelope via the docusign api? We already have the docusign integration working for that, so it'd probably be easier.

**Comment by davisagli** [comment]
I figured out how to pass the info we need to the powerform (both prefilling fields like WP Name and youth's name, and passing the youth's contact id as a hidden variable so we can map it to Salesforce).

This is a better solution than using the API because:
1. We can start the signing process even if we don't yet know the identity of the parent, and it'll prompt them for their name and email.
2. We don't have to get production access to the API, which requires a process to be certified as a Docusign developer partner.

I'll work on switching things to use the powerform approach.

**Comment by davisagli** [comment]
This should generate working powerform urls in the email reminder now. @mscholtz I need some tips on how to test the reminders in order to make sure it's working.

**Comment by skleinfeldt** [comment]
Assigning to @mscholtz to answer David's question.

**Comment by mscholtz** [comment]
@davisagli you would follow the testing instructions for story 49 (https://www.scrumdo.com/projects/project/wta/iteration/128090#story_803680) to trigger the reminder, and then just try a registration for a WP that also requires parental release
